pkg/virt-controller/watch/vm.go

```diff
@@ -58,104 +58,108 @@ func NewVMController(vmService services.VMService, recorder record.EventRecorder
+		handleSchedulingError(logger, err, queue, key)
 
-			// TODO move defaulting to virt-api
-			if vmCopy.Spec.Domain == nil {
-				spec := v1.NewMinimalDomainSpec(vmCopy.GetObjectMeta().GetName())
-				vmCopy.Spec.Domain = spec
+	} else {
+		logger.Info().Msg("Handing over the VM to the scheduler succeeded.")
+	}
+
+}
+func handleSchedulingError(logger *logging.FilteredLogger, err error, queue workqueue.RateLimitingInterface, key interface{}) {
+	logger.Error().Reason(err).Msg("Updating the VM state to 'Scheduling' failed.")
+	if e, ok := err.(*errors.StatusError); ok {
+		if e.Status().Reason == metav1.StatusReasonNotFound ||
+			e.Status().Reason == metav1.StatusReasonConflict {
+			// Nothing to do for us, VM got either deleted in the meantime or a newer version is enqueued already
+			return
+		}
+	}
+	queue.AddRateLimited(key)
+}
+func handleStartVMError(logger *logging.FilteredLogger, err error, vmService services.VMService, vmCopy v1.VM) {
+	logger.Error().Reason(err).Msg("Defining a target pod for the VM.")
+	pl, err := vmService.GetRunningPods(&vmCopy)
+	if err != nil {
+		logger.Error().Reason(err).Msg("Getting all running Pods for the VM failed.")
+		return
+	}
+	for _, p := range pl.Items {
+		if p.GetObjectMeta().GetLabels()["kubevirt.io/vmUID"] == string(vmCopy.GetObjectMeta().GetUID()) {
+			// Pod from incomplete initialization detected, cleaning up
+			logger.Error().Msgf("Found orphan pod with name '%s' for VM.", p.GetName())
+			err = vmService.DeleteVM(&vmCopy)
+			if err != nil {
+				logger.Critical().Reason(err).Msgf("Deleting orphaned pod with name '%s' for VM failed.", p.GetName())
+				break
+			}
+		} else {
+			// TODO virt-api should make sure this does not happen. For now don't ask and clean up.
+			// Pod from old VM object detected,
+			logger.Error().Msgf("Found orphan pod with name '%s' for deleted VM.", p.GetName())
+			err = vmService.DeleteVM(&vmCopy)
+			if err != nil {
+				logger.Critical().Reason(err).Msgf("Deleting orphaned pod with name '%s' for VM failed.", p.GetName())
+				break
 			}
-			vmCopy.Spec.Domain.UUID = string(vmCopy.GetObjectMeta().GetUID())
-			vmCopy.Spec.Domain.Devices.Emulator = "/usr/local/bin/qemu-x86_64"
-			vmCopy.Spec.Domain.Name = vmCopy.GetObjectMeta().GetName()
+		}
+	}
 
-			// TODO when we move this to virt-api, we have to block that they are set on POST or changed on PUT
-			graphics := vm.Spec.Domain.Devices.Graphics
-			for i, _ := range graphics {
-				if strings.ToLower(graphics[i].Type) == "spice" {
-					graphics[i].Port = int32(4000) + int32(i)
-					graphics[i].Listen = v1.Listen{
-						Address: "0.0.0.0",
-						Type:    "address",
-					}
+}
+func copyVM(vm *v1.VM) v1.VM {
 
-				}
+	vmCopy := v1.VM{}
+	model.Copy(&vmCopy, vm)
+	if vmCopy.Spec.Domain == nil {
+		spec := v1.NewMinimalDomainSpec(vmCopy.GetObjectMeta().GetName())
+		vmCopy.Spec.Domain = spec
+	}
+	vmCopy.Spec.Domain.UUID = string(vmCopy.GetObjectMeta().GetUID())
```

Comment:
Created #124 for this one. This and the next two lines should definitely not be here ...